ompi/mca/hook/base/hook_base.c

```diff
@@ -0,0 +1,299 @@
+        if( cli->cli_component == (mca_base_component_t*)comp ) {
+            found = true;
+        }
+    }
+
+    if( !found ) {
+        cli = OBJ_NEW(mca_base_component_list_item_t);
+        cli->cli_component = (mca_base_component_t*)comp;
+        opal_list_append(additional_callback_components, (opal_list_item_t*) cli);
+    }
+
+    return OMPI_SUCCESS;
+}
+
+int ompi_hook_base_deregister_callbacks(ompi_hook_base_component_t *comp)
+{
+    mca_base_component_list_item_t *cli, *found_cli = NULL;
+
+    // Check if it is already there
+    OPAL_LIST_FOREACH(cli, additional_callback_components, mca_base_component_list_item_t) {
+        if( cli->cli_component == (mca_base_component_t*)comp ) {
+            found_cli = cli;
+        }
+    }
+
+    if( NULL != found_cli ) {
+        opal_list_remove_item(additional_callback_components, (opal_list_item_t*) found_cli);
+        OBJ_RELEASE(found_cli);
+    }
+
+    return OMPI_SUCCESS;
+}
+
+MCA_BASE_FRAMEWORK_DECLARE(ompi, hook, "hook hooks",
+                           ompi_hook_base_register,
+                           ompi_hook_base_open,
+                           ompi_hook_base_close,
+                           mca_hook_base_static_components, 0);
+
+
+/* ***********************************************************************
+ * ***********************************************************************
+ * *********************************************************************** */
+
+/*
+ * If the framework has not been opened, then we can only use the static components.
+ *
+ * Otherwise we would need to initialize opal outside of ompi_mpi_init and possibly
+ * after ompi_mpi_finalize which gets messy (especially when tring to cleanup).
+ */
+#define HOOK_CALL_COMMON_NOT_IN_MPI(fn_name, ...)                       \
+    do {                                                                \
+        ompi_hook_base_component_t *component;                          \
+        int idx;                                                        \
+                                                                        \
+        for(idx = 0; NULL != mca_hook_base_static_components[idx]; ++idx ) { \
+            component = (ompi_hook_base_component_t*)mca_hook_base_static_components[idx]; \
+            if( NULL != component->hookm_ ## fn_name && ompi_hook_base_ ## fn_name != component->hookm_ ## fn_name ) { \
+                component->hookm_ ## fn_name ( __VA_ARGS__ );           \
+            }                                                           \
+        }                                                               \
+    } while(0)
+
+/*
+ * Once the framework is open then call all available components with
+ * the approprate function pointer. Call order:
+ * 1) static components
+ * 2) dynamic components
+ * 3) 'registered' components (those registered by ompi_hook_base_register_callbacks)
+ */
+#define HOOK_CALL_COMMON_IN_MPI(fn_name, ...)                           \
```

Comment:
This name (and HOOK_CALL_COMMON_NOT_IN_MPI) is misleading.  You call this macro (or the other one) depending on whether the hook framework has been initialized -- not whether we are in MPI or not.

This may be nit-picky, but it also could apply to OSHMEM someday.